### ISSUE TITLE
Lots of unk generated when translating with right-2-left decoding.

### ISSUE THREAD

**Issue by torshie** [pytorch/fairseq#1568]
I trained two en-de transformer models with some back-translation data. One is left-2-right decoding based, the other is right-2-left decoding based.

The L2R one works as expected, on WMT14, it obtained 29.7 bleu. However the R2L one generates lots of `<unk>` tokens. The R2L one can only got 28.0 bleu if those `<unk>` tokens weren't suppressed (`<unk>` can be suppressed with --unkpen 10.0 option during decoding.)

Why the R2L model generates so many `<unk>` tokens, while the L2R one generates none ?

#### Code
For the R2L one, I trained with the following code:

    import torch
    import fairseq.tasks.translation
    import fairseq.data
    
    
    class R2LDataset(fairseq.data.BaseWrapperDataset):
        def __getitem__(self, item):
            tensor = self.dataset[item]
            assert len(tensor.shape) == 1
            assert tensor[-1].item() == 2  # </s> symbol
    
            result = torch.empty_like(tensor)
            result[:-1] = torch.flip(tensor[:-1], (0,))
            result[-1] = tensor[-1]
            return result
    
    
    @fairseq.tasks.register_task('r2l_translation')
    class R2LTranslation(fairseq.tasks.translation.TranslationTask):
        def load_dataset(self, split, epoch=0, combine=False, **kwargs):
            super().load_dataset(split, epoch=epoch, combine=combine, **kwargs)
            original = self.datasets[split]
            assert type(original) == fairseq.data.LanguagePairDataset
            original.tgt = R2LDataset(original.tgt)

Training command line options are like:

    fairseq-train  data \
    --arch transformer_wmt_en_de_big \
    --share-all-embeddings \
    --optimizer adam \
    --adam-betas '(0.9,0.98)' \
    --clip-norm 0.0 \
    --lr 0.0005 \
    --lr-scheduler inverse_sqrt \
    --warmup-updates 4000 \
    --warmup-init-lr 1e-07 \
    --dropout 0.3 --weight-decay 0.0 \
    --criterion label_smoothed_cross_entropy \
    --label-smoothing 0.1 \
    --max-tokens 5000 \
    --attention-dropout 0.1 \
    --update-freq 10 \
    --max-update 41000 \
    --tensorboard-logdir /cache/nmt/tensorboard \
    --save-dir /cache/nmt/model \
    --fp16 \
    --log-format simple \
    --log-interval 1 \
    --user-dir ./code \
    --task r2l_translation

#### What's your environment?
The training node has 8 x Tesla v100 GPUs

**Comment by lematt1991** [comment]
Sorry, nothing immediately comes to mind as to why you'd be getting so many `unk`s for the R2L direction.  Maybe @myleott has thoughts?
